fs.go

```diff
@@ -0,0 +1,197 @@
+package dalec
+
+import (
+	"context"
+	"io"
+	"io/fs"
+	"path"
+	"sync"
+
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/tonistiigi/fsutil/types"
+
+	"github.com/tonistiigi/fsutil"
+
+	"github.com/moby/buildkit/client/llb"
+)
+
+var _ fs.DirEntry = &stateRefDirEntry{}
+var _ fs.ReadDirFS = &StateRefFS{}
+
+type StateRefFS struct {
+	s       llb.State
+	ctx     context.Context
+	opts    []llb.ConstraintsOpt
+	client  gwclient.Client
+	res     *gwclient.Result
+	o       sync.Once
+	initErr error
+}
+
+func NewStateRefFS(s llb.State, ctx context.Context, client gwclient.Client) *StateRefFS {
+	return &StateRefFS{
+		s:      s,
+		ctx:    ctx,
+		client: client,
+	}
+}
+
+func (fs *StateRefFS) fetchRef() (*gwclient.Result, error) {
+	def, err := fs.s.Marshal(fs.ctx, fs.opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := fs.client.Solve(fs.ctx, gwclient.SolveRequest{
+		Definition: def.ToPB(),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (fs *StateRefFS) Res() (*gwclient.Result, error) {
+	fs.o.Do(func() {
+		res, err := fs.fetchRef()
+		fs.res = res
+		fs.initErr = err
+	})
+
+	return fs.res, fs.initErr
+}
+
+func (fs *StateRefFS) Ref() (gwclient.Reference, error) {
+	res, err := fs.Res()
```

Comment:
We should be able to cache this.

Reply:
it should be cached already. The initialization is in a sync.Once, so all future calls after the first one should simply return `fs.res` without doing the initialization